docs/infra/module-dependencies.md

```diff
@@ -90,3 +90,129 @@ parent -- depends on --> config_value
 data.parent -- depends on --> config_value
 child -- depends on --> data.parent
 ```
+
+### Create "data" modules to reference a group of resources created in another module
+
+Often, a root module B will depend on a group of related resources created in another root module A. Moreover, the resources in A are typically created by calling a reusable module in [infra/modules](/infra/modules/) rather than directly in the root module. In this situation, rather than create a separate config for each dependency, create a data module D that B can call to reference R's resources, and group D and R together in a folder e.g. `/infra/modules/component/resources` and `/infra/modules/component/data`. In order to DRY up how related resources are named, create an static `interface` module that both the `resources` module and `data` module can reference. The `interface` module should typically have only one variable as input, the `name` of the group of resources, and the outputs should be the names of related resources in the group.
```

Comment:
My high level abstract POV on this pattern is that, while its very well designed, its very hard to get your head around. I don't think the average mid level engineer would be able to successfully replicate / expand on it in practice. I would look at this and go "...why...?" and then toss all the extra complexity (read: well structured design and patterns) out.

I'm not totally sure about the best way to address that, though...

Reply:
I agree that it's more complicated than I'd like and I generally am against complicated stuff if there are simpler patterns that achieve the same goals. In this case though I just haven't seen Terraform patterns that work effectively. One pattern I see for dependencies across root modules is to use Terraform's remote state data source, but I think that's more brittle and less explicit than what we have here since the module being depended on has no way of knowing what other modules depend on its remote state. So in this case I think I'm willing to eat the cost of inventing a new pattern that people will have to learn for the benefit of that pattern.

Reply:
Lets just hope it works out 🤞🏼